### ISSUE TITLE
Cannot reuse ConsulClientConfiguration for multiple clients

### ISSUE THREAD

**Issue by Eibwen** [PlayFab/consuldotnet#53]
The following throws an exception of `Cannot access a disposed object.Object name: 'System.Net.Http.WebRequestHandler'`.

This is because `ConsoleClient.Dispose` also calls down to `Config.Handler.Dispose()` and Config only creates the `Handler` in the constructor.

```
void Main()
{
	var config = new ConsulClientConfiguration();

	using (var client = new ConsulClient(config))
	{
		Console.WriteLine(client.Catalog.Services().Result.StatusCode);
	}
	using (var client = new ConsulClient(config))
	{
		Console.WriteLine(client.Catalog.Services().Result.StatusCode);
	}
}
```

I would expect the Configuration settings should not be a holder for a WebRequestHandler, or should rebuild it if the config object is reused.  OR the constructor of the client should take in a `Action<ConsulClientConfiguration>` to show it can't simply be reused.

**Comment by highlyunavailable** [comment]
Thanks for reporting this!

The reason the handler is in the `ConsulClientConfiguration` is because it's using the handler to do SSL certificate check ignoring and/or credentials passing based on the config values set.

The following statement does not excuse this bug, but: ideally you'd only make a single Client per configuration and reuse it for multiple requests all over your program since each request is self-contained/thread safe. That said, it is entirely expected that you _should_ be able to reuse the config object between multiple clients.

I'll take a look once I get home tonight and see if I can figure out a way around this that makes sense. I don't really want to move the "create handler" code into the ConsulClient because the behavior right now is to update the handler directly if you edit the Configuration, though technically that would be unsafe now that I think about it since there's no locks around it and it's supposed to be sort of a create-only kind of object.

Despite that, I'm thinking the cleanest way around this _would_ be to save the variables when set instead of directly updating the Handler, then create a new HttpHandler for each Client every time the config is passed to a ConsulClient's constructor. This has the side effect of "freezing" the config per client, so if you change the Config object's values, it won't change the value of the Client that's using it (e.g. changing authentication means you have to recreate all clients using that Config object) but that's a "safer" side effect than trying to manage shared state between clients, which would actually require locking of every config value - something that I don't particularly want to do, since it adds overhead to accessing the Config object.

If you have any further thoughts on what would be the "least surprising" way around this, given what I wrote above, I'll happily read them.

**Comment by Eibwen** [comment]
The second idea (first listed here) I alluded to would look like this, MassTransit is the main project where i've encountered this pattern, which allows for more flexibility for that since it supports multiple "wrapped" message queue services.

I think both of these are different than the ideas you suggest.

	//Changes to consul client:
	public ConsulClient()
	{
		//Normal
	}
	public ConsulClient(Action<ConsulClientConfiguration> configurator)
	{
		//The client controls the actual instanciation, not the consuming code
		Config = configurator(new ConsulClientConfiguration());
	}
	
	//Usage of this:
	using (var client = new ConsulClient(config => config.Addres = new Uri("http://mydockermachineip:8500"))
	{
		Console.WriteLine(client.Catalog.Services().Result.StatusCode);
	}

The other main idea i can think of is more or less:

```
public class ConsulClientConfiguration
{
	WebRequestHandler _handler;
	public WebRequestHandler Handler
	{
		get
		{
			if (_handler == null || _handler.IsDisposed()) //Can you check if its disposed easily, idk?
			{
				_handler = new WebRequestHandler();
				AddClientCertificateToHandler();
			}
			return _handler;
		}
	}
}
```

The second method here shouldn't change the public interface, but does make a lot more things you need to keep track of.  

The first method here might not be familiar for some people (likely requiring some docs around it).

Those were the ideas I thought of... totally up to you of course :)

**Comment by highlyunavailable** [comment]
You can't easily tell if a handler is disposed - you'd have to wrap `try {} catch (ObjectDisposedException) {RebuildHandlerAndHttpClient()}` around every use of it, which means every use of the HttpClient in the `ConsulClient` class. I am not sure how the `Action<ConsulClient>` version prevents someone from just passing a variable into the closure and thus bypassing it entirely by sharing state, and I _really_ do not want to change the public interface at this point. I'll give it some thought - thanks for the examples.

**Comment by Eibwen** [comment]
Yeah, I believe with the `Action<Configurator>` method, you need to focus on only allowing objects that are threadsafe and can be shared being able to be set from it due to that issue.

**Comment by highlyunavailable** [comment]
This issue has been resolved and is available on Nuget as [0.6.4.3](https://www.nuget.org/packages/Consul/0.6.4.3). Thanks again for reporting it!

**Comment by highlyunavailable** [comment]
@Eibwen I believe I implemented a version of what you were describing in your first implementation in https://github.com/PlayFab/consuldotnet/commit/6db3fec382f40a7eea3fcd5d815257b52720da9c and will cut a release next week. I do want to obsolete config object sharing in 0.7.0 since it's thread-unsafe to modify the object after the ConsulClient consumes it. You'll still be able to share configs if needed via sharing the "setup Action" delegate and calling it against multiple clients, but this is just a heads-up since you so recently requested this so you don't end up getting locked into an "old" style of doing it.
